src/iterators/not_iterator.c

```diff
@@ -226,6 +226,74 @@ static IteratorStatus NI_SkipTo_Optimized(QueryIterator *base, t_docId docId) {
   return rc;
 }
 
+// Revalidate for NOT iterator - Non-optimized version.
+static ValidateStatus NI_Revalidate_NotOptimized(QueryIterator *base) {
+  NotIterator *ni = (NotIterator *)base;
+
+  // 1. Revalidate the child iterator
+  ValidateStatus child_status = ni->child->Revalidate(ni->child);
+
+  // 2. Handle child validation results
+  if (child_status == VALIDATE_ABORTED) {
+    // Free child and replace with empty iterator
+    // When child is aborted, NOT iterator becomes "NOT nothing" = everything
+    ni->child->Free(ni->child);
+    ni->child = IT_V2(NewEmptyIterator)();
+    // Continue processing - this doesn't invalidate our current position
+  }
+
+  // Now the child is either at EOF, OK or MOVED.
+  // if the child is at EOF or OK, we can return VALIDATE_OK.
+  // if the child is MOVED, it must have advanced beyond the iterator's lastDocId, so the current result is still valid in this case.
+  RS_LOG_ASSERT(child_status != VALIDATE_MOVED || ni->child->lastDocId > base->lastDocId, "Moved but still not beyond lastDocId");
+  return VALIDATE_OK;
+}
+
+// Revalidate for NOT iterator - Optimized version.
+static ValidateStatus NI_Revalidate_Optimized(QueryIterator *base) {
+  NotIterator *ni = (NotIterator *)base;
+
+  // 1. Revalidate the wildcard iterator first
+  ValidateStatus wcii_status = ni->wcii->Revalidate(ni->wcii);
+  if (wcii_status == VALIDATE_ABORTED) {
+    return VALIDATE_ABORTED; // If wildcard iterator is aborted, we must abort too
+  }
+
+  // 2. Revalidate the child iterator
+  ValidateStatus child_status = ni->child->Revalidate(ni->child);
+  if (child_status == VALIDATE_ABORTED) {
+    // Free child and replace with empty iterator
+    // When child is aborted, NOT iterator becomes "NOT nothing" = everything
+    ni->child->Free(ni->child);
+    ni->child = IT_V2(NewEmptyIterator)();
+  }
+
+  // 3. Branch on wildcard status
+  if (wcii_status == VALIDATE_OK) {
+    // Same logic as non-optimized case
+    RS_LOG_ASSERT(child_status != VALIDATE_MOVED || ni->child->lastDocId > base->lastDocId, "Moved but still not beyond lastDocId");
+    base->atEOF = ni->wcii->atEOF; // sync EOF state with wildcard iterator
```

Comment:
is it needed? if it was not on EOF before, can it change with a VALIDATE_OK result?

Reply:
Correction after PR was changed:
`VALIDATE_OK` means the iterator was not changed, and if it wasn't EOF before the revalidation, `current` is still valid and has the same docID. Specifically, the EOF status is not expected to change if `VALIDATE_OK` is returned